### ISSUE TITLE
Bad type inference from returning Promise w/ struct types

### ISSUE THREAD

**Issue by johnw42** [google/closure-compiler#827]
This code is rejected:

    /**
     * @return {!Promise<{member: string}>}
     */
    function foo() {
      return Promise.resolve({member: ''});
    };

This code compiles:

    /**
     * @return {!Promise<{member: string}>}
     */
    function bar() {
      /** @type {{member: string}} */
      var result = {member: ''};
      return Promise.resolve(result);
    };

Here's the error message:

     ## /usr/local/google/home/jrw/chrome_dev/src/bad.js:5: ERROR - inconsistent return type
     ## found   : Promise<{member: string}>
     ## required: Promise<{member: string}>
     ##   return Promise.resolve({member: ''});
     ##          ^

And here are the flags I'm using:

    --accept_const_keyword --jscomp_error=accessControls --jscomp_error=ambiguousFunctionDecl --jscomp_error=checkStructDictInheritance --jscomp_error=checkTypes --jscomp_error=checkVars --jscomp_error=constantProperty --jscomp_error=deprecated --jscomp_error=externsValidation --jscomp_error=globalThis --jscomp_error=invalidCasts --jscomp_error=missingProperties --jscomp_error=missingReturn --jscomp_error=nonStandardJsDocs --jscomp_error=suspiciousCode --jscomp_error=undefinedNames --jscomp_error=undefinedVars --jscomp_error=unknownDefines --jscomp_error=uselessCode --jscomp_error=visibility --language_in=ECMASCRIPT5_STRICT --summary_detail_level=3 --jscomp_error=reportUnknownTypes --jscomp_error=duplicate --jscomp_error=misplacedTypeAnnotation

**Comment by MatrixFrog** [comment]
Also repro's in the debugger: https://closure-compiler-debugger.appspot.com/#input0%3D%252F**%250A%2520*%2520%2540return%2520%257B!Promise%253C%257Bmember%253A%2520string%257D%253E%257D%250A%2520*%252F%250Afunction%2520foo()%2520%257B%250A%2520%2520return%2520Promise.resolve(%257Bmember%253A%2520''%257D)%253B%250A%257D%253B%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26CHECK_TYPES%3D1%26MISSING_PROPERTIES%3D1

**Comment by MatrixFrog** [comment]
With the new type inference, I get a different error:

ERROR - Could not instantiate type TYPE|IThenable<TYPE>|null|undefined with {member:string}.

We may not be able to fix this in the old type inference but it would be nice to at least have a sensible error message.

**Comment by MatrixFrog** [comment]
Added this to `RecordType`:

    public String toStringHelper(boolean forAnnotations) {
      String str = super.toStringHelper(forAnnotations);
      if (!forAnnotations) {
        str += (declared ? " (declared)" : "");
      }
      return str;
    }

and then you get

    found   : Promise<{member: string}>
    required: Promise<{member: string} (declared)>

so at least now we know why those two seemingly identical types are considered different.

**Comment by concavelenz** [comment]
I've seen a similar problem in the past and it resolved to a record type, and a union of record types that was basically:
(A|A) != A
And when it was printed it printed as "A" instead of "(A|A)".  Perhaps that is what is going on here?

**Comment by johnw42** [comment]
@MatrixFrog From my perspective as a (relatively) humble user, putting the string (declared) in the error message doesn't clear much up. All I really know is what I knew before, i.e. two types that look identical to me are considered different by the compiler.

Generally speaking, would it be advisable to switch my project to use the new type inferencer yet?

**Comment by dimvar** [comment]
> Generally speaking, would it be advisable to switch my project to use the new type inferencer yet?

No. The new type inference is not production-ready yet. There are many bugs to fix before it's ready. We will announce it on the mailing list when it's good enough for people to switch.

**Comment by Dominator008** [comment]
Looks like NTI doesn't give a warning any more.
